py/htm/encoders/eye.py

```diff
@@ -0,0 +1,627 @@
+        # Sanity checks.
+        assert(len(self.image.shape) == 3)
+        assert(self.image.shape[2] == 3) # Color images only.
+        self.reset()
+        self.center_view()
+
+    def center_view(self):
+        """Center the view over the image"""
+        self.orientation = 0
+        self.position    = (self.image.shape[0]/2., self.image.shape[1]/2.)
+        self.scale       = np.min(np.divide(self.image.shape[:2], self.retina_diameter))
+
+    def randomize_view(self, scale_range=None):
+        """Set the eye's view point to a random location"""
+        if scale_range is None:
+            scale_range = [2, min(self.image.shape[:2]) / self.retina_diameter]
+        assert(len(scale_range) == 2)
+        self.orientation = random.uniform(0, 2 * math.pi)
+        self.scale       = random.uniform(min(scale_range), max(scale_range))
+        roi_radius       = self.scale * self.retina_diameter / 2
+        self.position    = [random.uniform(roi_radius, dim - roi_radius)
+                                 for dim in self.image.shape[:2]]
+
+    def _crop_roi(self):
+        """
+        Crop to Region Of Interest (ROI) which contains the whole field of view.
+        Adds a black circular boarder to mask out areas which the eye can't see.
+
+        Note: size of the ROI is (eye.output_diameter * eye.resolution_factor).
+        Note: the circular boarder is actually a bit too far out, playing with
+          eye.fovea_scale can hide areas which this ROI image will show.
+
+        Arguments: eye.scale, eye.position, eye.image
+
+        Returns RGB image (diameter * diameter, but effectively cropped to an 
+          inner circle - FOV). 
+
+        See also, @see make_roi_pretty()
+        """
+        assert(self.image is not None)
+
+        r     = int(round(self.scale * self.retina_diameter / 2))
+        x, y  = self.position
+        x     = int(round(x))
+        y     = int(round(y))
+        x_max, y_max, color_depth = self.image.shape
+
+        # Find the boundary of the ROI and slice out the image.
+        x_low  = max(0, x-r)
+        x_high = min(x_max, x+r)
+        y_low  = max(0, y-r)
+        y_high = min(y_max, y+r)
+        image_slice = self.image[x_low : x_high, y_low : y_high]
+
+        # Make the ROI and insert the image into it.
+        roi = np.zeros((2*r, 2*r, 3,), dtype=np.uint8)
+        if x-r < 0:
+            x_offset = abs(x-r)
+        else:
+            x_offset = 0
+        if y-r < 0:
+            y_offset = abs(y-r)
+        else:
+            y_offset = 0
+        x_shape, y_shape, color_depth = image_slice.shape
+        roi[x_offset:x_offset+x_shape, y_offset:y_offset+y_shape] = image_slice
+
+        # Rescale the ROI to remove the scaling effect.
+        roi.resize( (self.retina_diameter, self.retina_diameter, 3))
+
+        # Mask out areas the eye can't see by drawing a circle boarder.
+        # this represents the "shape" of the sensor/eye (comment out to leave rectangural)
```

Comment:
ok to crop to circular region here (and not only in the visualization)? Makes encoder see only ROI as the inner circle.